### DISCUSSION TITLE
Write data to Service ROS is wrong

### DISCUSSION THREAD

**Discussion by Nhavilla** [eProsima/Fast-DDS#4453] [Unofficial support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

When I create a client service dds to write data to service Ros, the data transferred to ros is incorrect or cannot be transferred. Ros received the DDS client but no incoming data or wrong data between variables.
I can still publish to Ros topic with correct data

### Current behavior

I used fastdds -gen to convert from .idl .srv  file

### Steps to reproduce

These values are passed wrongly to other values, it is not correct with its name

### Fast DDS version/commit

FastDDS ver 2.7
ROS Galactic

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

Default configuration, UDPv4 & SHM

**Comment by jsan-rt** [comment]
Hello @Nhavilla, thanks for using Fast-DDS.

Could you provide us with a minimal reproducer and the .idl file you are using?

**Comment by Nhavilla** [comment]
Hello @jsantiago-eProsima, thank you.

This is the .idl service file I use

>  module robot_services {
    module srv {
    struct Savemap_Request {
      string name;
    };
    struct Savemap_Response {
      boolean saved;
    };
  };
}

and the code
>DdsRos::MapServiceRequest::MapServiceRequest()
    : participant_(nullptr)
    , publisher_(nullptr)
    , topic_(nullptr)
    , writer_(nullptr)
    , type_(new robot_services::srv::Savemap_RequestPubSubType())
{
}
DdsRos::MoveServiceRequest::~MoveServiceRequest() 
{
    if (writer_ != nullptr)
    {
        publisher_->delete_datawriter(writer_);
    }
    if (topic_ != nullptr)
    {
        participant_->delete_topic(topic_);
    }
    if (publisher_ != nullptr)
    {
        participant_->delete_publisher(publisher_);
    }
    DomainParticipantFactory::get_instance()->delete_participant(participant_);
}
bool DdsRos::MoveServiceRequest::init()
{   

    DomainParticipantQos participantQos;
    participantQos.name("Move_pubscriber_client");

    
    participant_ = DomainParticipantFactory::get_instance()->create_participant(39, participantQos);

    if (participant_ == nullptr)
    {
        return false;
    }

    // Register the Type
    type_.register_type(participant_);

    std::string dds_topic_pose="rq/savemapRequest";
    
    
    // Create the subscriptions Topic
    topic_ = participant_->create_topic(dds_topic_pose, type_.get_type_name(), TOPIC_QOS_DEFAULT);

    if (topic_ == nullptr)
    {
        return false;
    }

    // Create the Subscriber
    publisher_ = participant_->create_publisher(PUBLISHER_QOS_DEFAULT, nullptr);

    if (publisher_ == nullptr)
    {
        return false;
    }

    // Create the DataWriter
    writer_ = publisher_->create_datawriter(topic_, DATAWRITER_QOS_DEFAULT, &listener_);

    if (writer_ == nullptr)
    {
        return false;
    }

    return true;

}

void DdsRos::MoveServiceRequest::PubListener::on_publication_matched(
    DataWriter* ,
    const PublicationMatchedStatus& info) 
{
    if (info.current_count_change == 1)
    {   
        matched_ = info.total_count;
        std::cout << "Publisher matched MoveService "<<matched_<< std::endl;

    }
    else if (info.current_count_change == -1)
    {   
        matched_ = info.total_count;
        std::cout << "Publisher unmatched  MoveService" << std::endl;

    }
    else
    {
        std::cout << info.current_count_change
                  << " is not a valid value for SubscriptionMatchedStatus current count change" << std::endl;
    }
    
}

void DdsRos::MoveServiceRequest::run(uint32_t samples)

{
    robot_services::srv::Savemap_Request map;

    uint32_t samples_sent = 0;
    while (samples_sent < samples)
    {   std::this_thread::sleep_for(std::chrono::milliseconds(2000));
        // publish();
        if (listener_.matched_ > 0)
        {   
            
            map.name()="Movin6j";
            writer_->write(&map);
        }
        
        samples_sent++;

        std::cout<<"request pub: "<<mov.name()<<std::endl;

        // std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
}

**Comment by Mario-DL** [comment]
Since the environment used in this issue (`Fast DDS 2.7.x` and `ROS Galactic`) is not being maintained anymore, I am moving this issue to the corresponding discussion forum.
